### PR TITLE
pythonPackages.typeguard: init at 2.3.1

### PR DESCRIPTION
**Pull request by obadz**
###### Motivation for this change

Add typeguard. First time I contribute a python package so wanted to get another pair of eyes before I get too comfortable with merging these things.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by obadz** [comment]
Well looks like I had contributed 44157c6da4 too but that was eons ago…

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -23151,6 +23151,26 @@ in {
     };
   };
 
+  typeguard = buildPythonPackage rec {
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix]: Due the size of this file, we add all new python packages to python-modules and only reference them here.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -23151,6 +23151,26 @@ in {
...
+    name = "typeguard-${version}";
+    version = "2.1.3";
+
+    src = pkgs.fetchurl {
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix]: fetchpypi can be used here.

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (Mic92) [inline pkgs/top-level/python-packages.nix]: There are test, but they are probably not available on pypi. If this is the case we usually download from github instead.
They can be run by adding `pytest` to `buildInputs` and override checkPhase:

```
checkPhase = "py.test .";
```

**Review by Mic92 — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -23151,6 +23151,26 @@ in {
...
+      url = "mirror://pypi/t/typeguard/${name}.tar.gz";
+      sha256 = "0l3pih5ca469v7if255h5rqymirsw46bi6s7p885jxhq1gv6cfpk";
+    };
+
+    propagatedBuildInputs = with self; [
+      setuptools_scm
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: this is typically only a build-time dependency and not a run-time dependency.

**Comment by obadz** [comment]
Thanks all for the feedback. How does the latest commit look?

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -23178,7 +23180,6 @@ in {
   runsnakerun = buildPythonPackage rec {
     name = "runsnakerun-2.0.4";
 
-
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: note this unrelated change

**Comment by FRidh** [comment]
looks good to me

File: pkgs/top-level/python-packages.nix (hunk 4 above)
Comment (obadz) [inline pkgs/top-level/python-packages.nix]: yes there was an extra empty line…

**Comment by obadz** [comment]
By the way, after running the tests, I see this exception/stack trace but not sure why. `py.test` still succeeds though (returns 0):

```

**Comment by FRidh** [comment]
Yes, I noticed it as well. It seems like they're not using the weak references correctly. I suppose upstream already knows about this?

**Comment by obadz** [comment]
Posted upstream. Merging this for now.
